src/app.cpp

```diff
@@ -80,34 +82,113 @@ void operator delete[](void* data) noexcept {
+                exit(0);
+            }
+            else if (arg == "--output" || arg == "-o") {
+                if (i + 1 >= argc) {
+                    exitWithError(argv[0], "missing filename after --output");
+                }
+                cli.output = argv[++i];
+            }
+            else if (arg == "--include") {
+                if (i + 1 >= argc) {
+                    exitWithError(argv[0], "missing filename after --include");
+                }
+                cli.includes.push_back(argv[++i]);
+            }
+            else if (arg == "--stack") {
+                if (i + 1 >= argc) {
+                    exitWithError(argv[0], "missing value after --stack");
+                }
+                try {
+                    cli.stack = std::stoll(argv[++i]);
+                }
+                catch (...) {
+                    exitWithError(argv[0], "--stack: value is not an integer");
+                }
+            }
+            else if (arg == "--skip-preprocessor") {
+                cli.skipPreprocessor = true;
+            }
+            else if (arg == "--silent") {
+                cli.silent = true;
+            }
+            else {
+                std::cerr << "Warning: unrecognized flag: " << arg << std::endl;
+            }
+        }
+        else {
+            std::filesystem::path input(arg);
+            if (!std::filesystem::is_regular_file(input)) {
+                exitWithError(argv[0], "input file not found");
+            }
+            cli.inputFiles.push_back(arg);
+        }
+    }
+
+    return cli;
+}
+
+int App::exec(const Cli& cli) {
     try {
         using namespace std::chrono;
         auto beg = high_resolution_clock::now();
         bool wait = false;
 
-        for (int i = 1; i < argc; ++i) {
+        for (auto const& input : cli.inputFiles) {
             max_memory = 0;
 
-            std::filesystem::path path = argv[i];
-
-            InputStream inputStream(FileR::read(path));
+            InputStream inputStream(FileR::read(input));
 
             Tokenizer tokenizer;
             tokenizer.run(inputStream);
+            std::vector<Token> tokens = tokenizer.data();
 
-            Preprocessor preprocessor(tokenizer.data());
-            preprocessor.run(path);
-            if (preprocessor.failed()) {
-                wait = true;
-                continue;
+            std::filesystem::path outputFile = cli.output.value_or(std::filesystem::current_path() / "tileset.rules");
+
+            if (!cli.skipPreprocessor) {
+                Preprocessor preprocessor(std::move(tokens));
+                preprocessor.run(input);
+                if (preprocessor.failed()) {
+                    wait = true;
+                    continue;
+                }
+                tokens = preprocessor.data();
+                max_memory = preprocessor.stack();
+                outputFile = cli.output.value_or(preprocessor.path() / (preprocessor.tileset() + ".rules"));

```

Comment:
As this is pretty much a replacement for directives, I would prefer if the preprocessor grabbed cli and handled all of that logic. App::exec() shouldn't need to change.